### ISSUE TITLE
the mcp-k8s is always running and hanging the console , is the normal

### ISSUE THREAD

**Issue by mting806** [strowk/mcp-k8s-go#11]
hello
I tried to use it with mcp python client(stdio).
The client will be hanging after it gets printout.
It looks like mcp-k8s is hanging the console.
is it by design?
how to exit it?
thank you very much

**Comment by strowk** [comment]
Hi!

I didn't observe that behavior you describe before. Could you provide more details? Preferably in the form of the code, so that I could try it out for myself.

To be clear - the process is supposed to keep running so long as until you haven't sent it a signal to stop (standard SIGTERM should do), this is pretty much by MCP design AFAIK

**Comment by mting806** [comment]
hello
i install it by npm -g
~~~
/usr/local/lib
├── @strowk/mcp-k8s@0.0.21-beta.1
├── corepack@0.30.0
└── npm@10.9.2
~~~
direct run it(i click 3 times return) , it will be hanging
~~~
venvtest@test-MBP misc % mcp-k8s

{"jsonrpc":"2.0","error":{"code":-32700,"message":"Parse error","data":"unexpected end of JSON input"},"id":null}

{"jsonrpc":"2.0","error":{"code":-32700,"message":"Parse error","data":"unexpected end of JSON input"},"id":null}

{"jsonrpc":"2.0","error":{"code":-32700,"message":"Parse error","data":"unexpected end of JSON input"},"id":null}

~~~

my test codes, I have to add a kill function at the end
~~~

import os
import signal
import psutil
import json

from mcp import ClientSession, StdioServerParameters

from mcp.client.stdio import stdio_client
import asyncio

def kill_process_by_name(name):
    for proc in psutil.process_iter():
        if proc.name() == name:
            os.kill(proc.pid, signal.SIGINT)

server_params = StdioServerParameters(
    command="mcp-k8s",
    args=[],
    env=None
)

async def run():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            result = await session.call_tool(
                name="list-k8s-namespaces",
                arguments={
                    "context": "kr-02"
                }
                )
            for content in result.content:
                print(json.loads(content.text))

            kill_process_by_name("mcp-k8s-go")

if __name__ == "__main__":
    asyncio.run(run())

~~~

**Comment by strowk** [comment]
> direct run it(i click 3 times return) , it will be hanging

That is totally how it is supposed to happen. When you hit return, you are simply sending process the empty line, to which it responds to you with "Parse error" with code `-32700`, which is correct by this specification - https://www.jsonrpc.org/specification

You can read more about design of stdio transport here: https://spec.modelcontextprotocol.io/specification/basic/transports/#stdio

If you want to stop it from your terminal, you should send it SIGTERM (this is usually done by pressing Ctrl-C).

As to the Python approach, I am not 100 percent sure at the moment. Might depend on some subtleties of what exactly mcp package is doing there. I have an idea, but need to confirm this first.

**Comment by mting806** [comment]
this is a test server with python sdk, its hanging by direct executing , 
python -m server.py
but with client call , it will be exit normally
update client code like
~~~
server_params = StdioServerParameters(
    command="python",
    args=["-m", "mcp_test.server"],
#    command="mcp-k8s",
#    args=[],
    env=None
)
~~~

server.py
~~~

from mcp.server import Server, NotificationOptions
from mcp.server.models import InitializationOptions
import mcp.server.stdio
import mcp.types as types
import asyncio




server = Server("example-server")


@server.list_prompts()
async def handle_list_prompts() -> list[types.Prompt]:
    return [
        types.Prompt(
            name="example-prompt",
            description="An example prompt template",
            arguments=[
                types.PromptArgument(
                    name="arg1",
                    description="Example argument",
                    required=True
                )
            ]
        )
    ]


@server.get_prompt()
async def handle_get_prompt(
    name: str,
    arguments: dict[str, str] | None
) -> types.GetPromptResult:
    if name != "example-prompt":
        raise ValueError(f"Unknown prompt: {name}")

    return types.GetPromptResult(
        description="Example prompt",
        messages=[
            types.PromptMessage(
                role="user",
                content=types.TextContent(
                    type="text",
                    text="Example prompt text"
                )
            )
        ]
    )


async def run():
    async with mcp.server.stdio.stdio_server() as (read_stream, write_stream):
        await server.run(
            read_stream,
            write_stream,
            InitializationOptions(
                server_name="example",
                server_version="0.1.0",
                capabilities=server.get_capabilities(
                    notification_options=NotificationOptions(),
                    experimental_capabilities={},
                )
            )
        )



if __name__ == "__main__":
    asyncio.run(run())
~~~

**Comment by mting806** [comment]
looks the package missed file execution setting
I have to chmod +x the mcp-k8s-go to make its working.
But yes , the hanging console status is resolved.
~~~
node:child_process:921
    throw err;
    ^

<ref *1> Error: spawnSync /usr/local/lib/node_modules/@strowk/mcp-k8s/node_modules/@strowk/mcp-k8s-darwin-x64/bin/mcp-k8s-go EACCES
    at Object.spawnSync (node:internal/child_process:1120:20)
    at spawnSync (node:child_process:868:24)
    at Object.execFileSync (node:child_process:911:15)
    at Object.<anonymous> (/usr/local/lib/node_modules/@strowk/mcp-k8s/bin/cli:34:14)
    at Module._compile (node:internal/modules/cjs/loader:1740:14)
    at Object..js (node:internal/modules/cjs/loader:1905:10)
    at Module.load (node:internal/modules/cjs/loader:1474:32)
    at Function._load (node:internal/modules/cjs/loader:1286:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:234:24) {
  errno: -13,
  code: 'EACCES',
  syscall: 'spawnSync /usr/local/lib/node_modules/@strowk/mcp-k8s/node_modules/@strowk/mcp-k8s-darwin-x64/bin/mcp-k8s-go',
  path: '/usr/local/lib/node_modules/@strowk/mcp-k8s/node_modules/@strowk/mcp-k8s-darwin-x64/bin/mcp-k8s-go',
  spawnargs: [],
  error: [Circular *1],
  status: null,
  signal: null,
  output: null,
  pid: 0,
  stdout: undefined,
  stderr: undefined
}

Node.js v23.5.0
~~~

**Comment by strowk** [comment]
> looks the package missed file execution setting

Oops, yup, this is because I have messed up this version a tad, tried to get it out faster and simply built it on my windows machine, while totally forgetting that it is unable to do chmod. Thanks for noticing, next release would be on done from CI properly!

I am closing this, as the issue is fixed now, and I would release 0.0.21 now to correct my previous chmod screwup. Should be ready in a couple of minutes.
